### ISSUE TITLE
Segfault when setting usertype std::vector

### ISSUE THREAD

**Issue by Matt8898** [ThePhD/sol2#181]
This was tested with lua 5.3.3 and luajit 2.0.4
Code:
```
#include <sol.hpp>
#include <string>
#include <vector>

struct test {
	std::vector<std::string> strings;
};

int main() {
	sol::state state;
	state.open_libraries(sol::lib::base, sol::lib::package);
	state.new_usertype<test>("test",
				 "strings", &test::strings);
	test a;
	a.strings.push_back("a");
	state["a"] = a;
	state.script("t = test.new() \n t.strings = a.strings");
}
```
GDB output(compiled with -g3):
```
Program received signal SIGSEGV, Segmentation fault.
0x000000000040db7e in sol::stack::getter<std::vector<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >*, void>::get_no_nil (L=0x62f728, index=3, tracking=...) at /usr/include/sol.hpp:5127
5127                                    void* udata = *pudata;

```
Backtrace:
```
#0  0x000000000040c7da in sol::stack::getter<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*, void>::get_no_nil (L=0x62c488, index=3, 
    tracking=...) at /usr/include/sol.hpp:5127
#1  0x000000000040c422 in sol::stack::getter<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, void>::get (L=0x62c488, index=3, 
    tracking=...) at /usr/include/sol.hpp:5163
#2  0x000000000040c33f in sol::stack::getter<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, void>::get (L=0x62c488, index=3, 
    tracking=...) at /usr/include/sol.hpp:4807
#3  0x000000000040bfb8 in sol::stack::stack_detail::unchecked_get<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > (L=0x62c488, index=3, 
    tracking=...) at /usr/include/sol.hpp:3934
#4  0x000000000040c0ca in sol::stack::stack_detail::evaluator::eval<sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, , 0ul, , std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&> (L=0x62c488, start=3, tracking=..., fx=<unknown type in /home/matt/projects/lua/a.out, CU 0x0, DIE 0x1b2cc>, 
    fxargs#0=@0x62ed90: &test::strings, fxargs#1=...) at /usr/include/sol.hpp:6350
#5  0x000000000040c1bd in sol::stack::stack_detail::call<false, 0ul, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&>(sol::types<void>, sol::types<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::integer_sequence<unsigned long, 0ul>, lua_State*, int, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller&&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&) (ta=..., tai=..., L=0x62c488, start=3, fx=<unknown type in /home/matt/projects/lua/a.out, CU 0x0, DIE 0x1b2cc>, args#0=@0x62ed90: &test::strings, args#1=...) at /usr/include/sol.hpp:6371
#6  0x000000000040bebb in sol::stack::call<false, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&>(sol::types<void>, sol::types<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, lua_State*, int, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller&&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&) (tr=..., ta=..., L=0x62c488, start=3, 
    fx=<unknown type in /home/matt/projects/lua/a.out, CU 0x0, DIE 0x1b2cc>, args#0=@0x62ed90: &test::strings, args#1=...) at /usr/include/sol.hpp:6417
#7  0x000000000040bd11 in sol::stack::call_into_lua<false, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&>(sol::types<void>, sol::types<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, lua_State*, int, sol::wrapper<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, void>::caller&&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&, test&) (tr=..., ta=..., L=0x62c488, start=3, 
    fx=<unknown type in /home/matt/projects/lua/a.out, CU 0x0, DIE 0x1b2cc>, fxargs#0=@0x62ed90: &test::strings, fxargs#1=...) at /usr/include/sol.hpp:6437
#8  0x000000000040bad8 in sol::call_detail::lua_call_wrapper<test, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, false, true, false, 0, void>::call_assign<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&> (L=0x62c488, f=@0x62ed90: &test::strings, o=...)
    at /usr/include/sol.hpp:7177
#9  0x000000000040b6c8 in sol::call_detail::lua_call_wrapper<test, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, false, true, false, 0, void>::call_assign<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&> (L=0x62c488, f=@0x62ed90: &test::strings)
    at /usr/include/sol.hpp:7191
#10 0x000000000040b0a9 in sol::call_detail::lua_call_wrapper<test, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, false, true, false, 0, void>::call_const<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&> (L=0x62c488, args#0=@0x62ed90: &test::strings)
    at /usr/include/sol.hpp:7206
#11 0x000000000040aca5 in sol::call_detail::lua_call_wrapper<test, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, false, true, false, 0, void>::call<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&> (L=0x62c488, f=@0x62ed90: &test::strings)
    at /usr/include/sol.hpp:7216
#12 0x000000000040a75a in sol::call_detail::call_wrapped<test, false, true, 0, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*&> (
    L=0x62c488, fx=@0x62ed90: &test::strings) at /usr/include/sol.hpp:7447
#13 0x000000000040a256 in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::real_call_with<1ul, false, true>(lua_State*, sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>&) (L=0x62c488, um=warning: RTTI symbol not found for class 'sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>'
...) at /usr/include/sol.hpp:9491
#14 0x0000000000409d47 in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::real_find_call<0ul, 1ul, false>(std::integral_constant<bool, false>, lua_State*) (this=0x62ed78, L=0x62c488)
    at /usr/include/sol.hpp:9388
#15 0x00000000004093dd in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::find_call<0ul, 1ul, false>(std::integral_constant<bool, false>, lua_State*, bool&, int&, sol::string_detail::string_shim const&) (this=0x62ed78, idx=..., L=0x62c488, found=@0x7fffffffe07b: true, ret=@0x7fffffffe07c: 0, accessor=...) at /usr/include/sol.hpp:9403
#16 0x0000000000408d8b in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::propogating_call<false>(lua_State*, bool&, int&, sol::string_detail::string_shim&) (this=0x62ed78, L=0x62c488, 
    found=@0x7fffffffe07b: true, ret=@0x7fffffffe07c: 0, accessor=...) at /usr/include/sol.hpp:9408
#17 0x000000000040fa6d in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::core_indexing_call<false, true>(lua_State*) (L=0x62c488) at /usr/include/sol.hpp:9459
#18 0x000000000040ec42 in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::real_new_index_call(lua_State*) (L=0x62c488) at /usr/include/sol.hpp:9479
#19 0x000000000040ec5d in sol::detail::static_trampoline<&sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::real_new_index_call>(lua_State*) (L=0x62c488) at /usr/include/sol.hpp:2862
#20 0x000000000040e519 in sol::usertype_metatable<test, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, char const (&) [8], std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > test::*, char const (&) [4], sol::constructor_list<sol::types<> >&, char const (&) [5], sol::destructor_wrapper<void> const&>::new_index_call(lua_State*) (L=0x62c488) at /usr/include/sol.hpp:9509
#21 0x00007ffff7bacc6e in ?? () from /usr/lib/liblua.so.5.3
#22 0x00007ffff7bad053 in ?? () from /usr/lib/liblua.so.5.3
#23 0x00007ffff7bb78ec in ?? () from /usr/lib/liblua.so.5.3
#24 0x00007ffff7bba27b in ?? () from /usr/lib/liblua.so.5.3
#25 0x00007ffff7bad05f in ?? () from /usr/lib/liblua.so.5.3
#26 0x00007ffff7bad0b1 in ?? () from /usr/lib/liblua.so.5.3
#27 0x00007ffff7bac482 in ?? () from /usr/lib/liblua.so.5.3
#28 0x00007ffff7bad33d in ?? () from /usr/lib/liblua.so.5.3
#29 0x00007ffff7ba8ab1 in lua_pcallk () from /usr/lib/liblua.so.5.3
#30 0x0000000000404877 in sol::stack::script (L=0x62c488, code="t = test.new() \n t.strings = a.strings") at /usr/include/sol.hpp:6467
#31 0x0000000000404afd in sol::state_view::script (this=0x7fffffffe4e0, code="t = test.new() \n t.strings = a.strings") at /usr/include/sol.hpp:10662
#32 0x000000000040307f in main () at main.cpp:17
quit
```

**Comment by ThePhD** [comment]
**TLDR; don't have a fix for you yet, need to bikeshed about making a breaking change to be more consistent.**

Oh, boy. You've opened a can of worms here... Shit's messy, yo. I don't have a good answer for you right now. But, here's a problem exposed by the earlier design of `sol`, which I thought about but haven't changed since the Sol 1.0 days:

`std::vector`, `std::list`, basically anything with a `begin()` and `end()` get pushed as a Table to Lua from C++ using Sol. However, no specific getter for `std::vector` or its other `begin()` and `end()` iterable types was defined.

Why?

We'd have to create a new vector, everytime, in order to accomodate the "table-like" semantics of handling a `std::vector` (since it gets pushed as a Table into Lua). I don't want to do that because it will give people the wrong idea and the performance will be shit.

Someone long ago submitted a PR to preserve the semantics of C++ vector types (and apparently other types), but the original author didn't think we needed it (it was beyond a year ago) and at the time I didn't think Sol would need it either. It would fix this but would ultimately make `vector` and friends no longer gettable or behave as tables, which breaks a lot of OTHER people's code...

**Comment by ThePhD** [comment]
Bit the bullet. I just decided to go ahead and add the non-performant version. If people get antsy I'll be sure add docs describing ways around it.

Thanks for reporting this.

Fixed here: https://github.com/ThePhD/sol2/releases/tag/v2.11.4

**Comment by Matt8898** [comment]
Thanks for your help.
